### ISSUE TITLE
Is there a way to reduce the size of output pdf files?

### ISSUE THREAD

**Issue by jvmachados** [UB-Mannheim/zotero-ocr#42]
I tested this ocr tool on some PDFs I downloaded from Academia.edu and the results were great. However, there's a problem: it increased the file size by A LOT (ex: a 11.8 MB file turned a 107 MB pdf). 

I was hoping to use this tool to create searchable and conveniently highlightable PDFs using scans from physical books I have, but scanned files are normally huge on their own. When I ran zotero-ocr on one of my scans (257 MB) I ended up with a file that's over 2GB in size (it won't even open). :(

Is there something I can do to decrease the file sizes?

(I use Zotero 6.0.9 on Windows and have installed the latest version of zotero-ocr)

**Comment by mattiaTagliente** [comment]
I second this request.

**Comment by gazzar** [comment]
pdftoppm is producing pngs. Maybe if it could be swapped over to jpeg, perhaps as an option, it would shrink the pdf

**Comment by sojusnik** [comment]
I'm using https://github.com/ocrmypdf/OCRmyPDF as a manual workaround. Maybe it will be sufficient for your case.

**Comment by gazzar** [comment]
Back in June when I last tried this, I also resorted to using OCRmyPDF after trying zotero-ocr.
I used the Docker version and OCRed my pdf with the following command
docker run -i --rm jbarlow83/ocrmypdf - - <jp.pdf >jp_ocr.pdf -O3 --jobs 2

**Comment by zuphilip** [comment]
Yeah, the size can become quite large. Tesseract itself creates the PDF with the input we give it. Tesseract would also run on jpg images, but the quality of the OCR output also depends on the inputed images and the colors.

The `-O3` option form OCRmyPDF looks good and this tool also uses tesseract under the hood. Maybe, one could consider to switch the workflow to it...

**Comment by stweil** [comment]
Reducing the resolution like in pull request #41 would reduce the size a lot. Using JPEG 2000 files with lossy compression would allow really small PDF files. Ideally that should be implemented in Tesseract.

**Comment by aborel** [comment]
I think we can reasonably use JPEG as a pdftoppm output and tesseract input, preferably as a user-selectable option. A quick and dirty test looks promising.

**Comment by stweil** [comment]
We produce nearly all of our OCR from JPEG. I'd use JPEG instead of PNG if this works and is easy to implement. Too many options confuse the users. Therefore I'd not add a new one for switching between JPEG and PNG.

A future solution could eliminate pdftoppm and extract the image which is part of the original PDF. This would eliminate any format conversion and automatically get the image in the best quality.

**Comment by aborel** [comment]
I would definitely prefer getting rid of pdftoppm (as per the ongoing discussion in #80 ), but I still don't have a clue how to implement that so I'm looking for a pragmatic solution - which could of course be replaced by something better eventually. It's a simple change, as far as I can see.
Too many options leading to users confusion: understood. I think I will keep some hidden options for testing purposes, but indeed not everybody needs to figure out the fine-tuning of pdftoppm's JPEG output.

**Comment by aborel** [comment]
The tests on my local code yesterday have been quite successful. Playing around with the JPEG options of pdftoppm, I have found pretty good default values that reduce the image size by a factor of 2 to 4 compared to PNG without visible loss of quality at 300 dpi. In my samples (2 pdfs: 268 pages, black & white book scan + 4 pages, color commercial pamphlet), the size of the output PDFs is now roughly the same as the original, sometimes a bit smaller.

I'm counting this as a success :-) I still need to double-check a few things before I push the commit. @stweil would you like to review the code before we create a new release?

**Comment by stweil** [comment]
> would you like to review the code before we create a new release

If you create a pull request, I can review the changes there, and it will automatically be added to the release notes. I'd make the new release after the new code was merged.

**Comment by aborel** [comment]
The new 0.9.0 version published today should resolve the issue.

**Comment by aborel** [comment]
I'm withdrawing the release, there is a metadata problem somewhere - both Zotero 6 and Zotero 7.1beta (my development instance) claim that the plugin is incompatible... I must have done something wrong.

**Comment by stweil** [comment]
The original 0.9.0 was missing the information which tells clients about the latest release. Therefore for example my own local Zotero installation did not see an update. That's why I updated the 0.9.0 release.

Which plugin did you test, your local one or the last one from the 0.9.0 release?

**Comment by aborel** [comment]
On v7.1beta I started using my own, because the update wasn't coming (I understand now how it would never have come - my bad).
I didn't really touch v6 until later, I guess after your build - and now the plugin is disabled. It also got disabled on v7.1b, I assume after the update got properly distributed.

**Comment by aborel** [comment]
Thanks for the new release. It seems to install properly on Zotero 7.0.15.
It's disabled on my other versions:
- Zotero OCR is incompatible with Zoteto 6.0.37
- Zotero OCR is incompatible with Zotero 7.1-beta.23+6787dcd59.

**Comment by stweil** [comment]
Hm. I now tried 0.9.0 with Zotero 6.0.37 and 7.1-beta.23. Initially both complained, but now they work with the plugin.

**Comment by aborel** [comment]
I have reinstalled the plugin manually using the current XPI, it has been accepted by Zotero 6 and 7.1 - I suppose there was some incorrect cached data on my system. Things look OK now, thanks!

I still need an update of the correct build/release process in the README to avoid the same mistake in the future. Can you take care of that?

**Comment by aborel** [comment]
Today the plugin is enabled again on Z7.1beta but disables on 6.0.37.
Something's fishy but probably on my side. Users should really upgrade to Zotero 7 anyway :-) I'll see if we got more feedback in the coming days, and then I'll close the issue.

**Comment by tobiscode** [comment]
I had the same issue and found a solution, although it is very use-case-specific and hacky. I'm posting this here in case someone has the same situation, or if my approach is adaptable to more use cases.

Specifically, I used the OCR to add searchable text to a book PDF. The original PDF was not large, a couple of MB, but several hundred pages. This meant that while the OCR went well, the output was almost 200 MB. Simply making the output images smaller with ghostscript never achieved such a small file size, and looked terrible.

Using `pdfimages`, I found out that my original PDF used a compression called `jbig2` where each image (i.e., page of text) is composed of sub-images (i.e., one image for each letter). I had never heard of it, thought it was quite genius, and of course wanted to reuse this format. Comparing the `pdfimages` output of the original and OCRed file, it was possible to match each original image to an OCRed one. I found [this gist](https://gist.github.com/tecoholic/e44f727f77c74e50f4d70ff10539350b) that allowed me to understand how PDF image embedding works, and used it not to insert an arbitrary image, but the image from the original file. I then also had to re-add all the character-specific images into the OCRed file, but then it worked! My new file looks exactly like the old one (except the page numbering, which I'm guessing is from the zotero-ocr toolchain), has OCR text, and is only 1-2 MB larger than the original one. Perfect!

If anyone's interested, here's my (hacky, uncommented, unformatted, untyped - I know) code: https://gist.github.com/tobiscode/fb30f483cf21bd2a76fd0efbbff0134a - maybe it helps someone.

Cheers

**Comment by aborel** [comment]
The size problem is resolved already, we just left the issue open in case there were problems with upgrading.
Closing (I have identified the cause on my machine).
